### PR TITLE
Report Filtering Typeahead (FAD-5332)

### PR DESCRIPTION
**Pull request by jasonrhodes**
Lots going on here, highlights:

1. in `actions/summaryChart` we refresh the typeahead cache whenever the date range changes (which will include the first time when it changes from nothing to initial values) -- we don't need to update this cache at any other time (more on this later)

1. in `actions/reportFilters` there is the new `refreshTypeaheadCache` action creator that will dispatch a bunch of actions. This is where I thought we would need to chain, so the dispatches are wrapped in `Promise.all()`, the idea was I would but a `.then()` at the end there and then process all that data and dispatch another action to store the processed data. But on further consideration, a selector was the better choice to prevent duplicating large lists in the store.

1. `selectors/reportFilterTypeaheadCache` is a new folder and file, with a selector to pull from the various parts of the redux store where the typeahead cache values should live, process them to put them in the right format for the typeahead component, and return them. It gets used in the `mapStateToProps` function in `pages/reports/components/Filter.js`

1. The Typeahead component in `pages/reports/components/Typeahead.js` is using Downshift, a react lib that does all of the keyboard and click events to make the UI of the typeahead work. @jonambas worked on this in parallel to my data work here

Still to do: 
* [x] use the filters in the summary chart metrics request so we get the correct filtered data in the chart
* [x] figure out why the chart isn't actually working right now :sob:
* [x] fix the charts (once we figure out why they aren't working)

### CODE REVIEW

File: src/pages/reports/components/Typeahead.js [hunk 1]
```diff
@@ -1,75 +1,89 @@
...
         </div>
-        <TextField onChange={this.handleOnChange} />
-      </div>
-    );
+      );
+    }
+
+    // Downshift automatically sets inputValue to selected value (which is an object and it converts it to a string lol)
+    // So instead we manage input value manually.
+    const handleStageChange = ({ inputValue }) => {
```
Comment (jonambas) [inline src/pages/reports/components/Typeahead.js]: `handleSta(t)eChange`

typo :persevere:

Comment (jasonrhodes) [reply src/pages/reports/components/Typeahead.js]: hmm but how it work !

Comment (jonambas) [reply src/pages/reports/components/Typeahead.js]: Oh also here: `onStateChange={handleStageChange}`

File: src/pages/reports/components/Typeahead.js [hunk 2]
```diff
@@ -1,75 +1,89 @@
...
+      clearSelection
+    }) => {
 
-  render() {
-    const { options = []} = this.props;
+      const mappedMatches = matches.map((item, index) => ({
+        content: <Item value={item.value} helpText={item.type} />,
+        ...getItemProps({ item, index }),
+        highlighted: highlightedIndex === index,
+        className: classnames(selectedItem === item && styles.selected) // Styles does nothing, was testing className pass through
```
Comment (jonambas) [inline src/pages/reports/components/Typeahead.js]: Can delete this. Should we also filter out selected items from the list? I can use this class to do a `display: none` if thats easier

Comment (jasonrhodes) [reply src/pages/reports/components/Typeahead.js]: Oh good call, yeah display: none is probably a lot easier than removing/adding from that list

Comment (jonambas) [reply src/pages/reports/components/Typeahead.js]: actually `selectedItem` probably only points to the most recent selected item

File: src/pages/reports/components/Typeahead.js (hunk 1 above)
Comment (jasonrhodes) [reply src/pages/reports/components/Typeahead.js]: hahahahahaha 👌

File: src/pages/reports/components/Typeahead.js (hunk 2 above)
Comment (jasonrhodes) [reply src/pages/reports/components/Typeahead.js]: we can probably pass in a prop with selected filters and compare against that list? might be cool to do a style like "greyed out with a check mark" or something

Comment (jasonrhodes) [reply src/pages/reports/components/Typeahead.js]: or just hide idc

**Pull request merged**